### PR TITLE
installer: show tracebacks from builds to debug failed installs

### PR DESCRIPTION
**Pull request by tgamblin**
`installer.py` currently swallows the traceback and preserves only the error messaege if a build process fails.

- [x] preserve exceptions from failed build processes
- [x] print a full traceback for each one when running with `spack -d`

### CODE REVIEW

File: lib/spack/spack/installer.py [hunk 1]
```diff
@@ -2196,6 +2197,9 @@ def install(self) -> None:
         if failed_build_requests or missing:
             for _, pkg_id, err in failed_build_requests:
                 tty.error(f"{pkg_id}: {err}")
+                if spack.error.debug:
```
Comment (haampie) [inline lib/spack/spack/installer.py]:
```suggestion
                if spack.main.SHOW_BACKTRACE:
```

**Comment by haampie** [comment]
In the case of fail fast can the original exception be re-raised?

File: lib/spack/spack/installer.py (hunk 1 above)
Comment (tgamblin) [reply lib/spack/spack/installer.py]: done

**Comment by tgamblin** [comment]
@haampie 

> In the case of fail fast can the original exception be re-raised?

We already raise an `InstallError` from the `ChildError` before even getting to this code in the case of fail-fast.  We can't raise the original exception because we don't have it -- we have the `ChildError`, which was sent back from the child process that failed.

I was thinking we could add the original exception as the `__cause__` of the `ChildError`, but I would have to dig into that further. I am pretty sure there is no guarantee that all exceptions raised in a child process are serializable (which is why `ChildError` exists -- to proxy things with a serializable class). So preserving the child exception in the parent is probably not something we really want to do (unless you want certain failures to be *really* weird and hard to debug).

The ChildError is meant to do a decent job of proxying the original exception -- it has the type, message, and traceback -- so I think this mostly does what you want already.

**Comment by haampie** [comment]
> I am pretty sure there is no guarantee that all exceptions raised in a child process are serializable

I think it's generally supposed to work. Do you know any counterexamples? The only one I know is HTTPError, but urllib is configured to raise DetailedHTTPError instead which does not have this problem.

Anyways, not a blocker.

**Comment by tgamblin** [comment]
> I think it's generally supposed to work. Do you know any counterexamples?

I wouldn't want to make it something we just throw in the install error code path w/o a guarantee or at least more examination than I really want to do to get this traceback in.  See, e.g., https://github.com/PrefectHQ/prefect/issues/2178.

> Anyways, not a blocker.

ok if this looks good can you merge?

**Comment by haampie** [comment]
1. With `--fail-fast` the exception is printed twice. Once from this code in black/white with build process backtrace, and then once in color without the backtrace from the child process.

2. The backtrace in package.py is off due to our changes, so not sure how helpful it is. For example:

   ```
     File "/home/harmen/spack/var/spack/repos/builtin/packages/gmake/package.py", line 84, in install
    if spec.satisfies("@:4.2.1"):  # generated files in build dir
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   ```
   
   should've been line 83 which has `configure(f"--prefix={prefix}", *self.configure_args())`.

Also can you apply

```diff
diff --git a/lib/spack/spack/util/executable.py b/lib/spack/spack/util/executable.py
index 0c1901cb1a..d37228c6c4 100644
--- a/lib/spack/spack/util/executable.py
+++ b/lib/spack/spack/util/executable.py
@@ -259,7 +259,7 @@ def streamify(arg, mode):
             if " " in self.exe[0]:
                 message += "\nDid you mean to add a space to the command?"
 
-            raise ProcessError("%s: %s" % (self.exe[0], e.strerror), message)
+            raise ProcessError("%s: %s" % (self.exe[0], e.strerror), message) from e
 
         except subprocess.CalledProcessError as e:
             if fail_on_error:
@@ -267,7 +267,7 @@ def streamify(arg, mode):
                     str(e),
                     "\nExit status %d when invoking command: %s"
                     % (proc.returncode, cmd_line_string),
-                )
+                ) from e
         except subprocess.TimeoutExpired as te:
             proc.kill()
             out, err = proc.communicate()
```

cause the typical backtrace is hard to follow due to "during handling of this ... this happened".

**Comment by haampie** [comment]
Ok, so I think I'm confused with this:

https://github.com/spack/spack/blob/7ddb40a804cae75ba8755cdc5af48873049111b4/lib/spack/spack/build_environment.py#L1379-L1385

who is responsible for printing a backtrace now?
